arrow-libs/core/arrow-core-retrofit/src/main/kotlin/arrow/retrofit/adapter/either/networkhandling/NetworkEitherCallAdapter.kt

```diff
@@ -0,0 +1,87 @@
+package arrow.retrofit.adapter.either.networkhandling
+
+import arrow.core.Either
+import arrow.core.left
+import arrow.core.right
+import okhttp3.Request
+import okio.Timeout
+import retrofit2.Call
+import retrofit2.CallAdapter
+import retrofit2.Callback
+import retrofit2.Response
+import java.io.IOException
+import java.lang.reflect.Type
+
+internal class NetworkEitherCallAdapter<R>(
+  private val successType: Type
+) : CallAdapter<R, Call<Either<CallError, R?>>> {
+
+  override fun adapt(call: Call<R?>): Call<Either<CallError, R?>> = EitherCall(call, successType)
+
+  override fun responseType(): Type = successType
+}
+
+private class EitherCall<R>(
+  private val delegate: Call<R>,
+  private val successType: Type
+) : Call<Either<CallError, R>> {
+
+  override fun enqueue(callback: Callback<Either<CallError, R>>) = delegate.enqueue(
+    object : Callback<R> {
+
+      override fun onResponse(call: Call<R>, response: Response<R>) {
+        callback.onResponse(this@EitherCall, Response.success(response.toEither()))
+      }
+
+      private fun Response<R>.toEither(): Either<CallError, R> {
+        // Http error response (4xx - 5xx)
+        if (!isSuccessful) {
+          val errorBody = errorBody()?.string() ?: ""
+          return HttpError(code(), errorBody).left()
+        }
+
+        // Http success response with body
+        body()?.let { body -> return body.right() }
+
+        // if we defined Unit as success type it means we expected no response body
+        // e.g. in case of 204 No Content
+        return if (successType == Unit::class.java) {
+          @Suppress("UNCHECKED_CAST")
+          Unit.right() as Either<CallError, R>
+        } else {
+          UnexpectedCallError(
+            IllegalStateException(
+              "Response code is ${code()} but body is null.\n" +
+                "If you expect response body to be null then define your API method as returning Unit:\n" +
+                "@POST fun postSomething(): Either<CallError, Unit>"
+            )
+          ).left()
+        }
```

Comment:
I'm curious. What happens if I'd have `Person` as a `json` in `body` but it's optional.
So I'd like to have `suspend fun fetchOptionalPerson(id: UUID): Either<CallError, Person?>`, would that be possible in this case?

Reply:
No, unless your json parsing Converter (which you set when configuring your Retrofit instance) allows `null` bodies.
All 3 converters I tested throw an exception in this case. Maybe you can configure it for some converters, I didn't check.

Also it doesn't seem like a valid REST idiom to me. There are no "optional" resources. If a `Person` has been found, server will return a `Person`. If it hasn't been found, server should return "404 Not Found" or a similar error response.

Reply:
Right, that makes perfect sense! Thanks for the clear reply!

> Also it doesn't seem like a valid REST idiom to me.

Anything out of the 2xx range is not considered an `Either.Right` 👍